### ISSUE TITLE
Make all key properties have the GenerateValueOnAdd flag set by convention

### ISSUE THREAD

**Issue by ajcvickers** [aspnet/EntityFramework#1271]
We are making a change so that the default behavior for properties that are part if keys is to have value generation enabled. Several points:
* Even with this flag set values will only be generated if the value of the property has not been set (is the default) when the entity is added. This means that use of non-generated keys where values have been set should not be affected.
* There is chance of collision of real key space and temporary key space, but we think this is unlikely to be a problem in the wild. Where collisions might occur they can be avoided by always setting key values explicitly so value generation never kicks in or by configuring different value generators that do not collide.
* Always generating keys provides a more flexible experience for dealing with disconnected graphs since temporary values will be used which can then be changed before saving
* An exception will be thrown if key generation is needed but no key generator is configured This can be avoided by setting the values explicitly before add or adding a new value generator.
* An exception will be thrown if a temporary key value is still set when the state of the object is changed to Unchanged, Modified, or Deleted. This will happen if SaveChanges is given temp values and doesn't generate real values.

**Comment by ajcvickers** [comment]
A couple of notes: 
- This needs to be done for any property that is a principal key, not only for primary keys.
- We could consider changeing Property.DefaultGenerateValueOnAdd to:
```
protected virtual bool DefaultGenerateValueOnAdd => this.IsKey();
```
Or even to something more complicated that determines if this is the root principal in the chain. This is similar to what we do for IsReadOnly. However, it could be making the core metadata too magic/convention-based and hence we may not want to do it this way, and we may want to consider not doing IsReadOnly this way as well.

**Comment by ajcvickers** [comment]
Another note. Currently is a property has GenerateValueOnAdd set and is not marked as using a SQL Server sequence, then it will be assumed to be SQL Server Identity. This doesn't work because properties such as alternate principal keys need to have GenerateValueOnAdd set but should not be treated as SQL Server Identity by either Migrations or the update pipeline.

It seems like there should be a SQL Server-specific convention which determines which column, if any, should be an Identity column and annotate appropriately. Migrations should do nothing other than honor the annotation. The value generator selector should use a temporary generator for any property that is not marked as using a sequence. The update pipeline should do nothing for columns that are not annotated as Identity. If temp values still exist in columns other than the Identity column then the existing exception for this case is correct.

**Comment by smitpatel** [comment]
fixed via 5201798fe95d94e4c66f79f58ad4ad62a3ca878a
